Upgrade to GraalVM 20.3

gsmet opened the pull request:
Running CI in my fork to see how much fun we will have.

Fixes #13954

gsmet committed:
Upgrade to GraalVM 20.3

gsmet commented:
Two failures:

The Tika tests are failing with:
```
Detailed message:
Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
Trace: Object was reached by 
	reading field java.awt.image.ColorModel.colorSpace of
		constant java.awt.image.DirectColorModel@5c78ebab reached by 
	reading field java.awt.image.ColorModel.RGBdefault
Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
Trace: Object was reached by 
	reading field org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB.awtColorSpace of
		constant org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB@538958bf reached by 
	scanning method org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:290)
Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(Exception): 
	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:288)
	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.loadICCProfile(PDICCBased.java:266)
	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.<init>(PDICCBased.java:131)
	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.create(PDICCBased.java:160)
	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:192)
	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:70)
	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:53)
	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.getColorSpace(PDDeviceNProcess.java:75)
	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.toString(PDDeviceNProcess.java:103)
```

And also related to ImageIO:
```
Error:  io.quarkus.it.main.ImageIOITCase.testImageRead  Time elapsed: 0.13 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Response body doesn't match expectation.
Expected: "1x1"
  Actual: <!doctype html>
<html lang="en">
<head>
    <title>Internal Server Error - Caught exception during read: Error id d5cf123d-89fc-4890-9eb0-fe57fd546f98-1</title>
```

Strangely enough, we don't have a proper Exception in this case.

gsmet commented:
/cc @zakkak @galderz

zakkak commented:
In Mandrel this is fixed by https://github.com/graalvm/mandrel/pull/174

I tried patching Quarkus to work with GraalVM CE 20.3 as well but with no success.
You can see my patch here: https://github.com/zakkak/quarkus/commit/cf3f394a3effc82127b090f16557489d80cd0a9c
And the failure here: https://github.com/zakkak/quarkus/runs/1415733753#step:7:702

zakkak commented:
FTR: The commit that breaks the test is https://github.com/oracle/graal/commit/39d7ff72e9b02d24a333d176c9dc3166afabd6a3 (part of GraalVM CE 20.3) and the commit that fixes it is https://github.com/oracle/graal/commit/5d2997a9ac8cafdfc6b5c316532165839ad7d1af (part of Mandrel 20.3 and GraalVM 21.0 but not GraalVM 20.3).

gsmet commented:
@zakkak I don't understand your statement here:

> Unfortunately though, this results in the native image to rely on AWT at runtime

AFAICS, we already depended on AWT at runtime but, now, somehow the module is not available anymore?

At least that's my understanding of the error?

zakkak commented:
> @zakkak I don't understand your statement here:
> 
> > Unfortunately though, this results in the native image to rely on AWT at runtime

Sorry for that. What actually happens now (GraalVM 20.3 and Quarkus https://github.com/zakkak/quarkus/commit/cf3f394a3effc82127b090f16557489d80cd0a9c) that was not happening before (GraalVM 20.2 and Quarkus `master`) is the invocation of `System.loadLibrary("awt")` **at runtime** from `sun.java2d.cmm.lcms.LCMS$1.run(LCMS.java:208)`.

gsmet commented:
It did happen at build time before, then, right?

If so, maybe we should just call it at static init?

(throwing random ideas...)

zakkak commented:
> It did happen at build time before, then, right?

Yes and then because the corresponding class was initialized it didn't need to follow that path and load it during runtime.

> If so, maybe we should just call it at static init?
> 
> (throwing random ideas...)

Loading `libawt` doesn't help, calling something else that would prevent it from being loaded at runtime should work though, e.g., `CMSManager.getModule()`. 

However I am not able to verify it using:
```java
import java.awt.color.ColorSpace;
import sun.java2d.cmm.CMSManager;

public class ColorSpaceTest2 {
    private final static CMSManager.PCMM pcmm;

    static {
        pcmm = CMSManager.getModule();
    }

    public static void main(String[] args) {
        ColorSpace.getInstance(ColorSpace.CS_sRGB);
    }
}
```
and 
```
/opt/jvms/graalvm-ce-java11-20.3.0/bin/native-image --initialize-at-build-time= ColorSpaceTest2; ./colorspacetest2
```

It doesn't run the static initializer at build time, any hints on how to force it?

zakkak committed:
Avoid ICC_ColorSpace instances in Apache Tika to support GraalVM 20.3

zakkak commented:
@gsmet please have a look at #13718 for the Tika issue.

gsmet commented:
@zakkak cool, thanks! Once you have made the adjustments in the other PR, I will incorporate your commit here and see how it goes.

Any idea about the other failure? The one failing on finding the size of an image for which we don't have a proper exception?

gsmet commented:
I wonder if we have an issue with the error page in native. It's not the first time I don't see any useful exception there.

gsmet committed:
Move Tika substitutions to proper package

Also add a comment about removing the PDFBox one when migrating to
GraalVM 21.

gsmet commented:
@zakkak so the Integration Tests / Main fail with the exact same issue:
```
g.UnsatisfiedLinkError: no awt in java.library.path
	at com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibrary(NativeLibrarySupport.java:131)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:270)
	at java.lang.Runtime.loadLibrary0(Runtime.java:830)
	at java.lang.Runtime.loadLibrary(Runtime.java:232)
	at java.lang.System.loadLibrary(System.java:351)
	at sun.java2d.cmm.lcms.LCMS$1.run(LCMS.java:208)
	at java.security.AccessController.doPrivileged(AccessController.java:84)
	at sun.java2d.cmm.lcms.LCMS.getModule(LCMS.java:202)
	at sun.java2d.cmm.lcms.LcmsServiceProvider.getModule(LcmsServiceProvider.java:34)
	at sun.java2d.cmm.CMMServiceProvider.getColorManagementModule(CMMServiceProvider.java:31)
	at sun.java2d.cmm.CMSManager.getModule(CMSManager.java:68)
	at java.awt.color.ICC_Profile.getInstance(ICC_Profile.java:790)
	at java.awt.color.ICC_Profile.getInstance(ICC_Profile.java:1026)
	at java.awt.color.ICC_Profile.getInstance(ICC_Profile.java:991)
	at java.awt.color.ICC_Profile$2.run(ICC_Profile.java:940)
	at java.awt.color.ICC_Profile$2.run(ICC_Profile.java:936)
	at java.security.AccessController.doPrivileged(AccessController.java:84)
	at java.awt.color.ICC_Profile.getStandardProfile(ICC_Profile.java:935)
	at java.awt.color.ICC_Profile.getDeferredInstance(ICC_Profile.java:1067)
	at java.awt.color.ICC_Profile.getInstance(ICC_Profile.java:854)
	at java.awt.color.ColorSpace.getInstance(ColorSpace.java:298)
	at javax.imageio.ImageTypeSpecifier.&lt;clinit&gt;(ImageTypeSpecifier.java:73)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:351)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:271)
	at com.sun.imageio.plugins.png.PNGImageReader.getImageTypes(PNGImageReader.java:1703)
	at com.sun.imageio.plugins.png.PNGImageReader.readImage(PNGImageReader.java:1461)
	at com.sun.imageio.plugins.png.PNGImageReader.read(PNGImageReader.java:1810)
	at javax.imageio.ImageIO.read(ImageIO.java:1468)
	at javax.imageio.ImageIO.read(ImageIO.java:1420)
	at io.quarkus.it.corestuff.ImageIOSupport.doGet(ImageIOSupport.java:18)
```

What I don't understand is that it was working before so they broke AWT entirely in 20.3?

zakkak commented:
> @zakkak so the Integration Tests / Main fail with the exact same issue:

Yes that was expected. Sorry if I didn't make it clear. https://github.com/quarkusio/quarkus/pull/13718 fixes only the Apache Tika issue. I will work on the ImageIOTest case next.

> What I don't understand is that it was working before so they broke AWT entirely in 20.3?

The loading of `libawt` at runtime was not working in 20.2 as well. What changed is the set of classes that are allowed to get instantiated in the native-image heap. 20.3 is more strict than 20.2 causing the native image to try and instantiate at runtime some classes that try to load `libawt` (which is not supported). So the workaround is to try and either avoid instantiating such classes (both at compile and runtime time) or try to force instantiate the class causing the issue at compile time.

Those interested in more details can also check our chat with @gradinac in [slack](https://graalvm.slack.com/archives/CN9KSFB40/p1607097252223800)

zakkak commented:
> I wonder if we have an issue with the error page in native. It's not the first time I don't see any useful exception there.

How did you finally managed to see the exception you posted in https://github.com/quarkusio/quarkus/pull/13644#issuecomment-740122878?

gsmet commented:
The ugly way :).

The native tests are running in prod mode so you don't see the exceptions.

I went to `QuarkusErrorServlet` and forced `showStack` to true.

gsmet committed:
Disable Image IO tests for now

We will bring them back when upgrading to GraalVM 21 but for now they
don't work with GraalVM 20.3.

gsmet commented:
I disabled the ImageIO test.

gastaldi commented on integration-tests/main/src/main/java/io/quarkus/it/corestuff/ImageIOSupport.java:
```diff
@@ -14,9 +12,10 @@
-        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        BufferedImage image = ImageIO.read(classLoader.getResource("META-INF/resources/1px.png"));
-        resp.getWriter().write(image.getHeight() + "x" + image.getWidth());
+        // Please uncomment this when we upgrade from GraalVM 20.3 to GraalVM 21
+        //        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        //        BufferedImage image = ImageIO.read(classLoader.getResource("META-INF/resources/1px.png"));
+        //        resp.getWriter().write(image.getHeight() + "x" + image.getWidth());
```
Do you really need to leave this commented out? I thought the `@Disabled` below would be enough?

gsmet replied:
No, it won't. It will still be included in the native image and that's what causes the issue. The issue is at build time not runtime.

zakkak reviewed (changes requested):
I suggest removing the changes for Tika (and those from https://github.com/quarkusio/quarkus/pull/12912) and disabling both Tika and ImageIO till [20.3.1 gets out (Scheduled for Jan 19) ](https://www.graalvm.org/release-notes/version-roadmap/).

20.3.1 will feature the necessary changes to make both tests build and pass without the need of any work-arounds (see https://github.com/oracle/graal/issues/3077)

If 1.11.0.Final can be delayed a bit (currently targeted for Jan 16) we could make sure we have some time to test everything out with 20.3.1 before release

gsmet commented:
The core artifact will be built on January 13th and I don't think we will postpone the release.

But we can upgrade to 20.3.1 in a micro of the 1.11 branch.

As for the Tika workaround, let's keep it, I will revert it when upgrading to 20.3.1.
